p4_symbolic/symbolic/util.h

```diff
@@ -53,11 +54,49 @@ gutil::StatusOr<SymbolicTrace> MergeTracesOnCondition(
     const z3::expr &condition, const SymbolicTrace &true_trace,
     const SymbolicTrace &false_trace);
 
-// Transforms a value read from a TableEntry to a z3::expr.
-gutil::StatusOr<z3::expr> IrValueToZ3Expr(const pdpi::IrValue &value);
+// Transforms a hex string literal from bmv2 json to a pdpi::IrValue
+gutil::StatusOr<pdpi::IrValue> ParseIrValue(std::string value);
 
-// Transforms a string value from bmv2 json to a pdpi::IrValue
-gutil::StatusOr<pdpi::IrValue> StringToIrValue(std::string value);
+// Transforms a value read from bmv2 (e.g. hardcoded in the program)
+// to a z3::expr.
+// Essentially, this is just a formatting function that can format ipv4, ipv6
+// hexstrings, and macs as bitvectors in z3's format.
+// This function does not perform any string translation, and instead assumes
+// the values are already in the actual domain of values in the p4 program,
+// because p4 and bmv2 both do not support string translation, it is only used
+// as a logical translation at the boundry between P4RT and bmv2.
+gutil::StatusOr<z3::expr> Bmv2ValueZ3Expr(const pdpi::IrValue &value);
+
+// Transforms a value read from a table entry to a z3::expr.
+// On top of formatting ipv4, ipv6, hexstrings, and macs as bitvectors,
+// this function also translates string values to unique and arbitrary
+// bitvectors.
+// The mapping of strings to bitvectors is stored and used to map the same
+// string to the same consistent bitvector, and to also do a reverse-translation
+// when extracting values for headers and fields from the z3 model.
+gutil::StatusOr<z3::expr> P4RTValueZ3Expr(const std::string field_name,
+                                          const pdpi::IrValue &value);
+
+// Exposes a static mapping from string values to bitvector values.
+// We do not need to include field names here because we assume the string
+// values are unique.
+std::unordered_map<std::string, uint64_t> &StringToBitVectorTranslationMap();
```

Comment:
It's unusual that a mutable global variable is the right way to go. What's the reason for not passing around the mapping to make the dataflow more explicit?

Reply:
It kind of reminds me of the z3_context and how that one is global too. I think passing it around is fine. Probably the best way to go about it is to have this be a class, with these maps being private in it, and having the parsing/translation functions be part of it, and then this could be passed around as part of the SymbolicPerPacketState or next to it as a new parameter.

Reply:
I did not want to pass translation maps around explicitly through out the symbolic evaluation pipeline because they would be too distracting, they are only used in like two places. But I think organizing them as some class/struct, might make that concern mute.